### PR TITLE
Refactored CredentialsResolver

### PR DESCRIPTION
**Pull request by ciastektk**
| Question                                  | Answer
| ---------------------------------------- | ------------------
| **Type**                                   | improvement
| **Target Ibexa DXP version** | `v4.0.0`
| **BC breaks**                          | no
| **Doc needed**                       | no

This PR provides fixes for CredentialsResolver:

- refactored `ExportCredentialsResolver` and `CredentialsResolver`
- Changed `hasCredentials` method signature in CredentialsResolverInterface
- Fixed phpstan errors

#### Checklist:
- [x] Provided PR description.
- [x] Tested the solution manually.
- [ ] Provided automated test coverage.
- [x] Checked that target branch is set correctly (master for features, the oldest supported for bugs).
- [x] Ran PHP CS Fixer for new PHP code (use `$ composer fix-cs`).
- [x] Asked for a review (ping `@ezsystems/php-dev-team`).

### CODE REVIEW

File: src/lib/Config/CredentialsResolver.php [hunk 1]
```diff
@@ -12,29 +12,29 @@
...
     {
         $credentials = $this->getRequiredCredentials($siteAccess);
 
-        foreach ($credentials as $credentialKey => $credentialValue) {
-            if (empty($credentialValue)) {
-                return false;
+        $emptyCredentials = array_filter(
+            $credentials,
+            static function (?string $credential = null): bool {
+                return empty($credential);
```
Comment (adamwojs) [inline src/lib/Config/CredentialsResolver.php:34]: What is the reason to change original fail-fast approach ?

Comment (Steveb-p) [reply src/lib/Config/CredentialsResolver.php:34]: > What is the reason to change original fail-fast approach ?

Logic seems to be incorrect.

In the original code the method returned `false` when encountering an empty credential entry.

In the new code method returns `false` only if all entries are empty.

Comment (ciastektk) [reply src/lib/Config/CredentialsResolver.php:34]: > What is the reason to change original fail-fast approach ?

Sorry, for that. Before, I thought it was a better idea but you're right, it should be done by previous fail-fast approach.

File: src/lib/Config/CredentialsResolver.php [hunk 2]
```diff
@@ -12,25 +12,22 @@
...
     public function __construct(ConfigResolverInterface $configResolver)
     {
         $this->configResolver = $configResolver;
     }
 
-    abstract protected function getRequiredCredentials(?string $siteAccess = null): array;
-
     /**
-     * {@inheritdoc}
+     * @return array<string>
```
Comment (Steveb-p) [inline src/lib/Config/CredentialsResolver.php:23]:
```suggestion
     * @return array<string, string>
```

File: src/lib/Config/ExportCredentialsResolver.php [hunk 3]
```diff
@@ -15,34 +15,32 @@
...
     {
-        $this->method = $this->configResolver->getParameter(
-            'export.authentication.method',
-            Parameters::NAMESPACE,
-            $siteAccess
-        );
+        $requiredCredentials = $this->getRequiredCredentials($siteAccess);
 
-        if ($this->method === ExportMethod::USER && !$this->hasCredentials($siteAccess)) {
+        if ($requiredCredentials['method'] === ExportMethod::USER && !$this->hasCredentials($siteAccess)) {
```
Comment (Steveb-p) [inline src/lib/Config/ExportCredentialsResolver.php:22]: Is the `'method'` array entry is used somewhere else? If yes, can it be converted into a `const`, so the connection between usage here and in consuming class is more visible?

**Review by Steveb-p — APPROVED** [review]
To be completely frank I find the design of abstract `CredentialsResolver` rather confusing. Dev-user is supposed to supply `getRequiredCredentials()` method, and `hasCredentials()` return `false` only, if there is a falsy value in array.

Ideally the abstract variant should be dropped, or the special behavior of `hasCredentials()` explained - maybe in phpdoc.

PR title should probably be "Refactored CredentialsResolver", as there is no actual change in behavior of the class, only phpdoc declarations were added?

**Review by adamwojs — APPROVED** [review]

**Comment by ciastektk** [comment]
> To be completely frank I find the design of abstract `CredentialsResolver` rather confusing. Dev-user is supposed to supply `getRequiredCredentials()` method, and `hasCredentials()` return `false` only, if there is a falsy value in array.
> 
> Ideally the abstract variant should be dropped, or the special behavior of `hasCredentials()` explained - maybe in phpdoc.
> 
> PR title should probably be "Refactored CredentialsResolver", as there is no actual change in behavior of the class, only phpdoc declarations were added?

Added explanation here: https://github.com/ezsystems/ezrecommendation-client/pull/65/commits/976cc649c49903a35ee228487be48952f8b2c649#diff-841c647bbc4b2403fe8e900092a49f3b4d040d86767f0ebb64667e35a96e2907R28

**Comment by ciastektk** [comment]
@adamwojs, @Steveb-p, I added some changes here: https://github.com/ezsystems/ezrecommendation-client/pull/65/commits/976cc649c49903a35ee228487be48952f8b2c649 Could you please review it again?

File: src/lib/Client/EzRecommendationClientInterface.php [hunk 4]
```diff
@@ -24,14 +24,14 @@ interface EzRecommendationClientInterface
...
-    public function setCustomerId(int $customerId): self;
+    public function setCustomerId(?int $customerId = null): self;
 
     public function getCustomerId(): ?int;
 
     /**
      * @return \EzSystems\EzRecommendationClient\Client\EzRecommendationClientInterface
      */
-    public function setLicenseKey(string $licenseKey): self;
+    public function setLicenseKey(?string $licenseKey = null): self;
```
Comment (Steveb-p) [inline src/lib/Client/EzRecommendationClientInterface.php:34]: I noticed that `licenseKey` and `customerId` can now be null. In which situation is this valid?

If it's the case, then I assume that they are either both filled or both empty, so this method can be merged into one.

Second, I believe that providing a default value should not happen in this case. If `setLicense()` method is to be called with the intent of removing it, `null` value should be explicitly passed by the user.

Comment (ciastektk) [reply src/lib/Client/EzRecommendationClientInterface.php:34]: I changed this to be consistent with getters which signature are marked that they can return null: https://github.com/ezsystems/ezrecommendation-client/blob/master/src/lib/Client/EzRecommendationClientInterface.php#L27-L36. 

Answering on your question, `customerId` and `licenseKey` can be null e.g.: just right after install DXP or when customer doesn’t want to use the Personalization module. 

I’ld to open different PR where I change default settings for this bundle which set env's as optional. From 4.X customer can add more then one configurations and overwrite this settings by semantic configuration.:   

```yaml
parameters:
    ezrecommendation.default.authentication.customer_id: '%env(default::int:RECOMMENDATION_CUSTOMER_ID)%'
    ezrecommendation.default.authentication.license_key: '%env(default::RECOMMENDATION_LICENSE_KEY)%'
    ezrecommendation.default.included_content_types: []
    ezrecommendation.default.host_uri: '%env(default::RECOMMENDATION_HOST_URI)%'
```

**Review by adamwojs — APPROVED** [review]

**Pull request merged**